atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -142,19 +142,21 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
 
     @Override
     public TransactionAndImmutableTsLock setupRunTaskWithConditionThrowOnConflict(PreCommitCondition condition) {
-        LockImmutableTimestampResponse immutableTsResponse = timelockService.lockImmutableTimestamp(
-                LockImmutableTimestampRequest.create());
+        StartAtlasDbTransactionResponse transactionResponse = timelockService.startAtlasDbTransaction(
+                IdentifiedTimeLockRequest.create());
         try {
-            LockToken immutableTsLock = immutableTsResponse.getLock();
-            long immutableTs = immutableTsResponse.getImmutableTimestamp();
+            LockToken immutableTsLock = transactionResponse.immutableTimestamp().getLock();
+            long immutableTs = transactionResponse.immutableTimestamp().getImmutableTimestamp();
             recordImmutableTimestamp(immutableTs);
-            Supplier<Long> startTimestampSupplier = getStartTimestampSupplier();
+
+            cleaner.punch(transactionResponse.freshTimestamp());
+            Supplier<Long> startTimestampSupplier = Suppliers.ofInstance(transactionResponse.freshTimestamp());
```

Comment:
The ordering here looks good. Nothing really changed except that we now get the fresh timestamp and punch before the transaction even starts.

Reply:
Yep. It _does_ shift earlier, but I think that's fine. Punching is async and we punch again later; there aren't any significant steps we'll change the order of if I'm understanding this correctly.